### DISCUSSION TITLE
Golang backend?

### DISCUSSION THREAD

**Discussion by prologic** [socketsupply/socket#872] [Q&A]
Hi! 👋

Having read through many pages of the docs, it isn't super clear to me about whether I can write the backend in Go?

I have an existing Go library/application that is cryptography heavy and it would be great if it could be reused with minimal modifications.

Can I use SSC to build a cross platform app) Desktop and Mobile) and integrate my existing Go code somehow?

Also iOS is super important to me.

**Comment by jwerle** [comment]
hi @prologic!

It has been some time since I used Go, but will [`gomobile`](https://pkg.go.dev/golang.org/x/mobile/cmd/gomobile) work here with your project?

**Reply by prologic** [reply]
Depends what you mean by "work". Yes gomobile can build Go code as an Android AAR or iOS Framework to be bundled alongside the rest of your mobile app. But that's only half the picture here. The reason I'm asking the question in the first place is because the architecture behind Socket Runtime SDK is inriging, in that it seems pretty solid and would allow me to write mobile apps using familiar web technologies. The only thing I then need is an actual Go backend where I can reuse all of my existing code and use HTML/JS/CSS on the frontend.

**Reply by jwerle** [reply]
You should be able to do this on the desktop. Unfortunately, `fork` which `popen` requires, is forbidden on iOS so you will not be able to spawn a child process which is how back ends on the runtime work. It is possible to write a native extension for the runtime using our stable C ABI, so if possible, you could compile your go code as a statically linkable library that embeds into the runtime as an extension

**Reply by prologic** [reply]
Ahh I did not know that `fork()` was a forbidden syscall on iOS? (_doesn't it run a fork of Darwin anyway? 🤔_) -- Okay, so, my only option to use Socket Runtime, would be to write an extension. That doesn't sound so bad as Go is interoperable with C via CGO bindings.

I'm not that great at C myself, so could we possible get some CGO bindings of Socket Runtime's SDK going. I'd be happy to take this further and evaluate how feasible this would be from there 👌

**Reply by jwerle** [reply]
The C ABI should interop well with CGO. We do not have an example of this yet, but it should be possible!

**Comment by prologic** [comment]
(_starting a separate thread_) -- I also notice in the docs that it is possible to write extensions using Web Assembly. I can currently compile my Go application/library into a `.wasm` and we've built a fairly "okay-ish" web app this way reusing all the crypto code we write in Go. Could we go down this path? If so, an example how how would be nice 👌
